### ISSUE TITLE
Feature Request -- Theme level support for RSS feeds

### ISSUE THREAD

**Issue by architarious** [wvuweb/cleanslate-cms#150]
Currently, if you have a blog in slate (or any other site outside of cleanslate) the only way to pull content from it's RSS feed is to use the snippet in cleanslate.

This becomes problematic for the following reasons:
- If you have multiple sites that are pulling the same feed, you have to remember to change it in multiple places if the feed or blog changes/moves.
- it makes it harder to style things from one theme to another consistently, since it's not loaded in via a partial.
- The current feed snippet itself has little to no flexibility in ciphering content. You can only display a title and a link, there is currently no way to display when it was posted or any other meta data,... (outside of manipulating the DOM via jquery, which adds another layer of complexity, is time consuming, and would break if the feeds address ever changed)
- If you want to style this in hammer with the actual content instead of dummy content, you have to implement the snippet first in cleanslate, then copy the generated html, then paste it into your mock-data file. 

It would be ideal if we had a way to fix all of this by using radius tags to declare the url we want to pull from and the tags we want to filter or how they're going be formatted (similar to how we currently do with local cleanslate blogs)

Some of this may be do-able with a custom snippet, that also has a feature-request and  was logged earlier,.. either way though I figured I would throw it out there considering CPASS has a need for it and it would likely also be useful university wide if anyone wanted to make a pattern to display stuff like WVUToday posts.

It would also be great if we could use this for calendar events... so that we don't have to rely on two separate scripts (one to pull and one to format) like we currently do.

**Comment by adamjohnson** [comment]
#75 is related to this.

Not sure about the calendar events part. Might be tricky since that is an app from a vendor (we don't control stuff). @jja006 may be able to provide some insight in this regard.

**Comment by jasonmarkle** [comment]
:+1:  Hate to be finding this now, but it sort of messes up the plan for all the engineering department sites.

**Comment by nreckart** [comment]
**New tag:** `<r:xslt_transform url="http://example.com/blog.rss" [xslt_file="/path/to/file.xslt"]>`

Using `<r:xslt_transform>` you can retrieve XML from a _url_ and transform it into whatever you want via [XSLT](https://en.wikipedia.org/wiki/XSLT). The XSLT template can be supplied either in the contents of the `<r:xslt_transform>` tag or via the _xslt_file_ tag attribute, which would be a path to a file in the theme. The results are **cached for 15 minutes**, or until the page, site, or theme are modified.

**_Note:**_ If using an XSLT file, it should be stored in a directory other than the _views_ directory in the theme.

**Example 1: Customize WVUToday RSS feed**

In a theme template...

``` html
<r:xslt_transform url="http://feeds.feedburner.com/wvu/ilPI" xslt_file="/util/rss_1.xslt" />
```

In the '/util/rss_1.xslt' file within the theme...

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
<xsl:output method="html"/>
  <xsl:template match="/">
    <ol class="rss-items">
      <xsl:apply-templates select="//item[position() &lt; 4]"/>
    </ol>
  </xsl:template>

  <xsl:template match="item">
    <li>
      <a class="item-title">
        <xsl:attribute name="href">
          <xsl:value-of select="link"/>
        </xsl:attribute>
        <xsl:value-of select="title"/>
      </a>
      <div class="item-description">
        <xsl:value-of disable-output-escaping="yes" select="description"/>
      </div>
    </li>
  </xsl:template>
</xsl:stylesheet>
```

generates...

``` html
<ol class="rss-items">
  <li>
    <a href="http://feedproxy.google.com/~r/wvu/ilPI/~3/3VWL7hboNRM/wvu-ranked-among-elite-global-universities-in-regional-science-research" class="item-title">WVU ranked among elite global universities in regional science research</a><div class="item-description">West Virginia University is among the top universities in a global ranking of regional science research, demonstrating the institution's prominence in the field.<div class="feedflare">
    <a href="http://feeds.feedburner.com/~ff/wvu/ilPI?a=3VWL7hboNRM:xBioeGy2KYQ:yIl2AUoC8zA"><img border="0" src="http://feeds.feedburner.com/~ff/wvu/ilPI?d=yIl2AUoC8zA"></a> <a href="http://feeds.feedburner.com/~ff/wvu/ilPI?a=3VWL7hboNRM:xBioeGy2KYQ:F7zBnMyn0Lo"><img border="0" src="http://feeds.feedburner.com/~ff/wvu/ilPI?i=3VWL7hboNRM:xBioeGy2KYQ:F7zBnMyn0Lo"></a> <a href="http://feeds.feedburner.com/~ff/wvu/ilPI?a=3VWL7hboNRM:xBioeGy2KYQ:qj6IDK7rITs"><img border="0" src="http://feeds.feedburner.com/~ff/wvu/ilPI?d=qj6IDK7rITs"></a>
    </div>
      <img width="1" height="1" alt="" src="http://feeds.feedburner.com/~r/wvu/ilPI/~4/3VWL7hboNRM">
    </div>
  </li>
  <li>
    <a href="http://feedproxy.google.com/~r/wvu/ilPI/~3/2U9cGaczd0Y/wvu-hosts-first-regional-workshop-on-big-data-cybersecurity" class="item-title">WVU hosts first regional workshop on Big Data, cybersecurity</a><div class="item-description">A recent workshop, hosted by West Virginia University, brought together experts in Big Data and cybersecurity to address the research and workforce development needs of the region.<div class="feedflare">
    <a href="http://feeds.feedburner.com/~ff/wvu/ilPI?a=2U9cGaczd0Y:h8DQ7bhdTXA:yIl2AUoC8zA"><img border="0" src="http://feeds.feedburner.com/~ff/wvu/ilPI?d=yIl2AUoC8zA"></a> <a href="http://feeds.feedburner.com/~ff/wvu/ilPI?a=2U9cGaczd0Y:h8DQ7bhdTXA:F7zBnMyn0Lo"><img border="0" src="http://feeds.feedburner.com/~ff/wvu/ilPI?i=2U9cGaczd0Y:h8DQ7bhdTXA:F7zBnMyn0Lo"></a> <a href="http://feeds.feedburner.com/~ff/wvu/ilPI?a=2U9cGaczd0Y:h8DQ7bhdTXA:qj6IDK7rITs"><img border="0" src="http://feeds.feedburner.com/~ff/wvu/ilPI?d=qj6IDK7rITs"></a>
    </div>
      <img width="1" height="1" alt="" src="http://feeds.feedburner.com/~r/wvu/ilPI/~4/2U9cGaczd0Y">
    </div>
  </li>
  <li>
    <a href="http://feedproxy.google.com/~r/wvu/ilPI/~3/5vJHTrjz70c/wvu-law-students-publish-on-prestigious-oxford-human-rights-hub-blog" class="item-title">WVU Law students publish on prestigious Oxford Human Rights Hub Blog</a><div class="item-description">Articles by two West Virginia University College of Law students have been published on the Oxford Human Rights Hub Blog.<div class="feedflare">
    <a href="http://feeds.feedburner.com/~ff/wvu/ilPI?a=5vJHTrjz70c:_KQzZaKRSQs:yIl2AUoC8zA"><img border="0" src="http://feeds.feedburner.com/~ff/wvu/ilPI?d=yIl2AUoC8zA"></a> <a href="http://feeds.feedburner.com/~ff/wvu/ilPI?a=5vJHTrjz70c:_KQzZaKRSQs:F7zBnMyn0Lo"><img border="0" src="http://feeds.feedburner.com/~ff/wvu/ilPI?i=5vJHTrjz70c:_KQzZaKRSQs:F7zBnMyn0Lo"></a> <a href="http://feeds.feedburner.com/~ff/wvu/ilPI?a=5vJHTrjz70c:_KQzZaKRSQs:qj6IDK7rITs"><img border="0" src="http://feeds.feedburner.com/~ff/wvu/ilPI?d=qj6IDK7rITs"></a>
    </div>
      <img width="1" height="1" alt="" src="http://feeds.feedburner.com/~r/wvu/ilPI/~4/5vJHTrjz70c">
    </div>
  </li>
</ol>
```

**Example 2: Transform XML Yahoo Weather**

In a theme template...

``` html
<r:xslt_transform url="http://weather.yahooapis.com/forecastrss?w=26505">
  <?xml version="1.0" encoding="UTF-8"?>
  <xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
                  xmlns:yweather="http://xml.weather.yahoo.com/ns/rss/1.0" version="1.0">
    <xsl:output method="html"/>
    <xsl:template match="/rss/channel">
      <p>
        Sunrise <xsl:value-of select="yweather:astronomy/@sunrise"/>,
        Sunset <xsl:value-of select="yweather:astronomy/@sunset"/>
      </p>
      <xsl:value-of disable-output-escaping="yes" select="item/description"/>
    </xsl:template>
  </xsl:stylesheet>
</r:xslt_transform>
```

generates...

``` html
<p xmlns:yweather="http://xml.weather.yahoo.com/ns/rss/1.0">
        Sunrise 8:02 am,
        Sunset 3:39 pm</p>
<img src="http://l.yimg.com/a/i/us/we/52/29.gif"><br><b>Current Conditions:</b><br>
Partly Cloudy, 54 F<br><br><b>Forecast:</b><br>
Wed - Partly Cloudy. High: 56 Low: 52<br>
Thu - Mostly Cloudy. High: 57 Low: 49<br>
Fri - Mostly Cloudy. High: 54 Low: 50<br>
Sat - Partly Cloudy. High: 59 Low: 52<br>
Sun - Partly Cloudy. High: 56 Low: 43<br><br><a href="http://us.rd.yahoo.com/dailynews/rss/weather/Lammas__GB/*http://weather.yahoo.com/forecast/UKEN2484_f.html">Full Forecast at Yahoo! Weather</a><br><br>
(provided by <a href="http://www.weather.com">The Weather Channel</a>)<br>
```

Instead of reinventing the wheel and basically attempting to build an XML parser in Radius tags, it makes sense to utilize an industry standard that has been around forever, is well documented, and fits our needs perfectly.

I would like to use this same concept to possibly add support for applying a custom template for the RSS snippet and other CleanSlate snippets, but that'll be down the road.

**Comment by jasonmarkle** [comment]
After installing this in the [engineering-departments repo](https://stash.development.wvu.edu/projects/CST/repos/engineering-departments/browse/views/frontpage.html#21-23), I built on to it and tested it. As a raw URL, things work fine within the tag, but once I get a little more complex with passing it [custom site data](https://stash.development.wvu.edu/projects/CST/repos/engineering-departments/browse/config.yml#31) in a variable, it get's a little funky.

`<r:xslt_transform url="{$department_news_rss_url}" xslt_file="/util/rss_1.xslt" /`

It does at least get on screen, but it tosses an error message, "ERROR: Could not load the XML URL:"

I haven't tested with page data just because as I've built I've tried to sort of nestle the more difficult things that people could accidentally screw up out of the way. I figure if I can alleviate some headaches for later, the better.

Also, you can't parse the images from the current atom structure it doesn't seem like. [I'm still testing this as well though.](https://stash.development.wvu.edu/projects/CST/repos/engineering-departments/browse/util/rss_1.xslt#12-14)

**Comment by nreckart** [comment]
@jasonmarkle What URL are you using that is breaking?

**Comment by jasonmarkle** [comment]
@nreckart The variable being pulled in above is http://statler.sandbox.wvu.edu/news.rss?tags=mae

**Comment by nreckart** [comment]
Custom site data is accessed through the `<r:site:data>` tag, not the global variables namespace. Unfortunately, you can't use a tag as the attribute of another tag, so you'd need to basically define a template variable and assign the site value to it.

`<r:set_var name="url"><r:site:data name="department_news_rss_url"/></r:set_var>`

Then

`<r:xslt_transform url="{$url}" xslt_file="/util/rss_1.xslt" />`

Not the most elegant, but it should work. I'll look into possibly making custom site data available in the global namespace.

**Comment by jasonmarkle** [comment]
Yeah, that's what I was trying to do. I ended up creating a variable but I was calling it through the page data instead of site data which was giving me an error. That has been alleviated and I'm now working on trying to get the images and story link to pull over. They are currently output, but not actually displayed within the content of the page even though they are showing in the source. I think this is just an issue with the naming conventions not being pulled from the actual blog feed. Will follow up shortly if I can get things running on all ends.

**Comment by nreckart** [comment]
I replied to your RSS email. I'm trying to keep this issue clean and specific to the original issue. At this point I think the issues are restricted to your specific site implementation, so we should handle it outside of this issue.

**Comment by nreckart** [comment]
@jasonmarkle: I've made some changes and custom site data should now be accessible via the global variable namespace. So you should be able to access your site specific RSS URL as you originally intended, without needing to first reassign it to a template variable. 

`<r:xslt_transform url="{$department_news_rss_url}" xslt_file="/util/rss_1.xslt" />`

**Comment by adamjohnson** [comment]
In plain english, this means that, if you use [custom data](http://cleanslatecms.wvu.edu/how-to/theme-development/custom-data) (site or page), you can output that data via a variable in a tag _without_ having to use `<r:set_var />`.

Code example:

config.yml

``` yaml
custom_data:
  - key_1
```

frontpage.html

``` html
<a href="{$key_1}">The URL for this link comes from Custom Page Data via this page's properties</a>
```

---

If, for some (poor) reason, you have custom site data and custom page data with the same name (eg: `key_1`), custom page data will always override custom site data.
